Zend/zend_attributes.stub.php

```diff
@@ -6,3 +6,8 @@ final class Attribute
 {
     public function __construct(int $flags = Attribute::TARGET_ALL) {}
 }
+
+final class Deprecated
```

Comment:
Could you please elaborate what are the benefits of making it final?

Doesn't it make sense to inherit from this one? E.g. [PhpStorm's deprecated](https://github.com/JetBrains/phpstorm-attributes/blob/master/src/Deprecated.php) attribute.

Reply:
Good catch. For now the code uses an API that finds exact matches of Deprecated attribute usage, inheritance would require some changes. It should be possible to add them, i will look into it next time I work on the code.

Reply:
If it wasn't final, then the php compiler wouldn't be able to do this check by name and cache it in opcache without knowing what other files contained. Opcache works on a single file at a time - knowing what the parent classes of a given class are is impossible when you know only a single file's contents, so even making it emit E_DEPRECATED for a subclass would be impractical to do without hurting performance and require attempting to autoload any user-defined attribute when a file is being compiled.

It also makes it much easier to read code - you don't have to know the parent classes to tell if something would emit a deprecation notice at runtime.

-----

Speaking of which, it may be useful to prevent `class_alias` where the source is a built-in attribute (which has observable effects) because the aliased classes wouldn't actually have an effect, but that's out of scope.

Reply:
You are right, i forgot it will require autoloading during compile, which does not really work

Reply:
@TysonAndre, @beberlei, thanks for the clarification. I now understand the technical limitation.

I'm wondering if there is any other way to make the attributes compatible. Here are the ideas:
1. Match the signatures: `(string $message)` --> `(string $message, string $replacement)`
2. Add options: `(string $message)` --> `(string $message, ...$options)`
3. Split PhpStorm attribute into two separate: `#[Deprecated]` and `#[DeprecatedReplacement]`

While the latter is easier for this RFC, I believe such solution is not future-proof. I mean a similar issue may arise later with other attributes introduced and it might not be always possible to split.

What do you think?

/cc @nikic